lib/Frontend/ArgsToFrontendOutputsConverter.cpp

```diff
@@ -202,3 +217,209 @@ OutputFilesComputer::deriveOutputFileFromParts(StringRef dir,
+    : Args(args), Diags(diags), InputsAndOutputs(inputsAndOutputs),
+      OutputFiles(outputFiles), ModuleName(moduleName),
+      RequestedAction(
+          ArgsToFrontendOptionsConverter::determineRequestedAction(Args)) {}
+
+Optional<SupplementaryOutputPaths>
+SupplementaryOutputPathsComputer::computeOutputPaths() const {
+  Optional<std::vector<SupplementaryOutputPaths>> pathsFromUser =
+      getSupplementaryFilenamesFromArguments();
+  if (!pathsFromUser)
+    return None;
+
+  return computeOutputPathsForOneInput(
+      OutputFiles[0], (*pathsFromUser)[0],
+      InputsAndOutputs.firstInputProducingOutput());
+}
+
+Optional<std::vector<SupplementaryOutputPaths>>
+SupplementaryOutputPathsComputer::getSupplementaryFilenamesFromArguments()
+    const {
+
+  auto objCHeaderOutput =
+      readSupplementaryOutputArguments(options::OPT_emit_objc_header_path);
+  auto moduleOutput =
+      readSupplementaryOutputArguments(options::OPT_emit_module_path);
+  auto moduleDocOutput =
+      readSupplementaryOutputArguments(options::OPT_emit_module_doc_path);
+  auto dependenciesFile =
+      readSupplementaryOutputArguments(options::OPT_emit_dependencies_path);
+  auto referenceDependenciesFile = readSupplementaryOutputArguments(
+      options::OPT_emit_reference_dependencies_path);
+  auto serializedDiagnostics =
+      readSupplementaryOutputArguments(options::OPT_serialize_diagnostics_path);
+  auto loadedModuleTrace = readSupplementaryOutputArguments(
+      options::OPT_emit_loaded_module_trace_path);
+  auto TBD = readSupplementaryOutputArguments(options::OPT_emit_tbd_path);
+
+  if (!objCHeaderOutput || !moduleOutput || !moduleDocOutput ||
+      !dependenciesFile || !referenceDependenciesFile ||
+      !serializedDiagnostics || !loadedModuleTrace || !TBD) {
+    return None;
+  }
+  std::vector<SupplementaryOutputPaths> result;
+
+  const unsigned N =
+      InputsAndOutputs.countOfFilesProducingSupplementaryOutput();
+  for (unsigned i = 0; i < N; ++i) {
+    result.push_back(SupplementaryOutputPaths(
+        (*objCHeaderOutput)[i], (*moduleOutput)[i], (*moduleDocOutput)[i],
+        (*dependenciesFile)[i], (*referenceDependenciesFile)[i],
+        (*serializedDiagnostics)[i], (*loadedModuleTrace)[i], (*TBD)[i]));
+  }
+  return result;
+}
+
+// Extend this routine for filelists if/when we have them.
+
+Optional<std::vector<std::string>>
+SupplementaryOutputPathsComputer::readSupplementaryOutputArguments(
+    swift::options::ID path_id) const {
+  std::vector<std::string> paths = Args.getAllArgValues(path_id);
+
+  const unsigned N =
+      InputsAndOutputs.countOfFilesProducingSupplementaryOutput();
+  if (paths.size() > N) {
+    std::unique_ptr<llvm::opt::OptTable> Table = createSwiftOptTable();
+    Diags.diagnose(SourceLoc(), diag::error_too_many_supplementary_outputs,
+                   Table->getOptionName(path_id));
+    return None;
+  }
+  while (paths.size() < N)
+    paths.push_back(std::string());
```

Comment:
This doesn't make sense; like with `-o`, there should always be exactly the right number of paths, or 0 paths.

Reply:
(Honestly it *is* useful to be able to override previous options by having the last argument win, but that's still not what's going on here.)

Reply:
Do you mean that the code should emit a diagnostic and bail out if there are paths, but not the right number?

Reply:
I think so, yes.

Reply:
OK, done, I think.